---
title: "Dataset generation notebook for comparison study"
output: html_notebook
---

# Install and setup


filename = '../data/S050_CCRCC_Clark_Cell2019_SIRCLE/cpg_DE_stage1_sircle.csv'
annotation_file = '../data/S050_CCRCC_Clark_Cell2019/TCGA/tcga_hsapiens_gene_ensembl-GRCh38.p13.csv'
f = Csv(filename, 'chr', 'start', 'start', 'logFC', ["AveExpr","t","P.Value","adj.P.Val","B","cpg_id"],
           gene_start=4, gene_end=5, gene_chr=3,
           gene_direction=6, gene_name=2)
# Add the gene annot
f.set_annotation_from_file(annotation_file)
# Now we can run the assign values
f.assign_locations_to_genes()
f.save_loc_to_csv(f'../data/S050_CCRCC_Clark_Cell2019_SIRCLE/cpg_DE_stage1_annot_sircle.csv')
```{r}
packages <- c("tidyverse", "reticulate", "dplyr", "caret")
install.packages(setdiff(packages, rownames(installed.packages())))

library(caret)
library(tidyverse)
library(dplyr)
library(reticulate)
  
condaEnvName <- "ml"
envName <- NULL
envPath <- NULL
setupEnv <- F

## ------------ Setup the environment ----------- ##
if (! is.null(condaEnvName)) {
    use_condaenv(condaEnvName, required = TRUE)
    setupEnv = T
}
if (! is.null(envName)) {
    use_virtualenv(envName, required = TRUE)
    setupEnv = T
}
if (! is.null(envPath)) {
    use_python(envPath, required = TRUE)
    setupEnv = T
}
if (! setupEnv) {
    print("WARNING: NEXT TIME USE envNAME --> we are creating you a new virtual environment called sircle. THIS TAKES TIME, so next time pass envName=sircle to skip this step. ")
    print("WARNING: YOUR SYSTEM NEED PYTHON > 3.6. If it doesn't nothing will work, please update to python 3.6 or consider installing conda!")
    virtualenv_create(
      envname = "ml",
      python = NULL,
      packages = "scie2g",
      system_site_packages = getOption("reticulate.virtualenv.system_site_packages",
                                       default = FALSE)
    )
    use_python(envPath, required = TRUE)
    setupEnv = T
}

```

## Annotate Methylation CpGs to genes
```{r}
scie2g <<- import("scie2g")    # Make global

# Drop rows with a * in the chr column
mapping_df <- read.csv('illuminaMethyl450_hg38_GDC.tsv', sep="\t")
mapping_df <- mapping_df[mapping_df$chrom != "*", ]
write.csv(mapping_df, 'normal_chrs_illuminaMethyl450_hg38_GDC.csv')
## Assign methylation CpGs to gene identifiers (using the mapping provided)
# Here we assign CpGs if they fall within 2500 of the TSS of the gene.
gene_mapper <- scie2g$Csv('normal_chrs_illuminaMethyl450_hg38_GDC.csv', 'chrom', 
                     'chromStart', 'chromEnd', 'chromStart', c("id","gene"),
                     buffer_after_tss=as.integer(0),
                     buffer_before_tss=as.integer(2500),
                     buffer_gene_overlap=as.integer(0),
                     gene_start=as.integer(4), gene_end=as.integer(5), gene_chr=as.integer(3),
                     gene_direction=as.integer(6), gene_name=as.integer(2))
gene_mapper$set_annotation_from_file("tcga_hsapiens_gene_ensembl-GRCh38.p13.csv")
# Now we can run the assign values
gene_mapper$assign_locations_to_genes()
# Save the mapping to CSV file.
gene_mapper$save_loc_to_csv('illuminaMethyl450_hg38_GDC_2500_promoter_mapping.csv')
```

## Function for generating the datasets
```{r}
library(dplyr)
library(matrixStats)

## Annotate the methylation CpGs to gene names (using ensembl gene ID)
cpg_annotation <- read.csv('illuminaMethyl450_hg38_GDC_2500_promoter_mapping.csv')
cpg_annotation <- cpg_annotation[!is.na(cpg_annotation$external_gene_name), ]
# If there are CpGs that map to multiple genes only pick the gene that it is closest to
cpg_annotation$dist_to_gene <- as.numeric(cpg_annotation$start_position) - as.numeric(cpg_annotation$location_start) #chromStart is the position of the CpG start and start_position is the start of the TSS of the gene
rev_trans <- cpg_annotation[cpg_annotation$dist_to_gene < 0,]
rev_trans$dist_to_gene <- rev_trans$location_end - rev_trans$end_position
cpg_annotation$dist_to_gene[cpg_annotation$dist_to_gene < 0] <- rev_trans$dist_to_gene # i.e. it is reverse stranded
cpg_annotation <- cpg_annotation[order(cpg_annotation$id, cpg_annotation$dist_to_gene), ]
cpg_annotation <- cpg_annotation[!duplicated(cpg_annotation$id),] # Remove any duplicated CpGs (i.e. this is where a CpG mapped to multiple genes)
rownames(cpg_annotation) <- cpg_annotation$id

make_datasets <- function(data_name, data_folder) {
      message("Making datasets for CNV, RNAseq, and DNA methylation of ", data_name, " in folder: ", data_folder)
      ### reading copy number values and seperating them into amplification and deletion calls
      phenotype <- read.csv(gzfile(file.path(data_folder, paste0("TCGA-", data_name, ".GDC_phenotype.tsv.gz"))),  header = T, sep = "\t")
      # Print out the number of normal and tumour samples
      print.data.frame(as.data.frame(table(phenotype$sample_type.samples)))
      # Replace the '-' with '.' since this is what R will do anyway
      phenotype$submitter_id_safe <- gsub("-", ".",  phenotype$submitter_id.samples, fixed = TRUE)
      
      # Create a column that has the sample type and sample name merged
      phenotype$sample_id <- paste(gsub(" ", "", phenotype$sample_type.samples, fixed = TRUE), phenotype$submitter_id_safe, sep="_")
      
      
      if (data_name == 'SKCM') {
        # Then we want to only keep primary tumour and metastatic
        cond1 <- phenotype[phenotype$sample_type.samples == 'Metastatic', ] 
        cond2 <- phenotype[phenotype$sample_type.samples == 'Primary Tumor', ]
      } else {
      # We only want to keep normal and primary tumour
        cond1 <- phenotype[phenotype$sample_type.samples == 'Primary Tumor', ] 
        cond2 <- phenotype[phenotype$sample_type.samples == 'Solid Tissue Normal', ]
      }
      cond1$condition_id <- 1
      cond2$condition_id <- 0
      # Combine the two dataframes
      samples <- rbind(as.data.frame(cond1), as.data.frame(cond2))
      rownames(samples) <- samples$submitter_id_safe
      # Now we have a sample dataframe that is ordered by condition 1 followed by condition 2
      # We also have a column name that has informative information regarding the tumour type.
      # We'll use this new column to rename the other datasets so we can keep track of our samples more easily.
      
      ### reading copy number values and seperating them into amplification and deletion calls
      cn <- read.table(gzfile(file.path(data_folder, paste0("TCGA-", data_name, ".gistic.tsv.gz"))), header = T, sep = "\t")
      cn[1:5,1:5]
      rownames(cn) <- cn[,1]
      cn <- cn[,-1]
      
      ### reading gene expression file
      ge.exp <- read.table(gzfile(file.path(data_folder, paste0("TCGA-", data_name, ".htseq_counts.tsv.gz"))), header = T, sep = "\t")
      ge.exp[1:5,1:5]
      rownames(ge.exp) <- ge.exp[,1]
      # Set Ensembl ID to be the main ID
      ge.exp <- ge.exp[,2:length(colnames(ge.exp))]
      # Remove the normalisation by log2 + 1
      ge.exp <- (2^ge.exp) - 1 
      
      ## reading methylation data
      me.exp <- read.table(gzfile(file.path(data_folder, paste0("TCGA-", data_name, ".methylation450.tsv.gz"))), header = T, sep = "\t")
      me.exp[1:5,1:5]
      me.exp <- me.exp[-1,]
      rownames(me.exp) <- me.exp[,1]
      me.exp <- me.exp[,-1]
      me.exp <- na.omit(me.exp)
      
      # map between the two CPG annotation and the CPGs to get the gene IDs
      annotated_cpgs <- intersect(rownames(me.exp), cpg_annotation$id)
      me.exp_annot <- me.exp[annotated_cpgs,]
      mapping <- cpg_annotation[annotated_cpgs,]
      me.exp_annot$var <- rowVars(as.matrix(me.exp_annot))
      me.exp_annot$id <- mapping$ensembl_gene_id
      
      # Here we need to pick the CpG mapped to the gene with the highest variability
      me.exp_annot <- me.exp_annot[order(me.exp_annot$id, -abs(me.exp_annot$var) ), ]
      me.exp_annot <- me.exp_annot[ !duplicated(me.exp_annot$id), ]
      
      # Now finally consolidate this 
      me.exp <- me.exp[rownames(me.exp_annot),]
      mapping <- mapping[rownames(me.exp_annot), ]
      rownames(me.exp) <- me.exp_annot$id
      # Also make mapping fit with this (i.e. having the ensembl gene ID mapping)
      mapping <- mapping[mapping$ensembl_gene_id %in% rownames(me.exp),]
      
      rownames(mapping) <- mapping$ensembl_gene_id
      rownames(me.exp_annot) <- mapping$ensembl_gene_id
      
      # Now we just want to get the intersect of the rownames, first we merge this 
      ## identifying common genes and samples in all the datasets
      # We also want to make sure we have the phenotype
      # reorder the columns to just be those in our phenotype
      cols <- intersect(samples$submitter_id_safe, colnames(cn))
      a <- intersect(cols, colnames(ge.exp))
      a1 <- intersect(colnames(me.exp), a)
      
      # Remove the ensembl ID version from the rownames
      ens_id_sep <- data.frame(do.call('rbind', strsplit(as.character(rownames(ge.exp)),'.',fixed=TRUE)))
      rownames(ge.exp) <- ens_id_sep$X1
      
      ens_id_sep <- data.frame(do.call('rbind', strsplit(as.character(rownames(cn)),'.',fixed=TRUE)))
      rownames(cn) <- ens_id_sep$X1
      b <- intersect(rownames(cn), rownames(ge.exp))
      b1 <- intersect(rownames(me.exp), b)
      
      samples_ge <- intersect(samples$submitter_id_safe, colnames(ge.exp))
      ge.exp2 <- ge.exp[b1, samples_ge]
      message("Gene expression samples:")
      print.data.frame(as.data.frame(table(samples[samples_ge, ]$sample_type.samples)))
      
      samples_cn <- intersect(samples$submitter_id_safe, colnames(cn))
      cn2 <- cn[b1, samples_cn]
      message("Copy number variation samples:")
      print.data.frame(as.data.frame(table(samples[samples_cn, ]$sample_type.samples)))
      
      samples_me <- intersect(samples$submitter_id_safe, colnames(me.exp))
      me.exp2 <- me.exp[b1, samples_me]
      message("DNA Methylation samples:")
      print.data.frame(as.data.frame(table(samples[samples_me, ]$sample_type.samples)))
      
      # Rename each of them to have more informative column names based on the sample name and sample type
      colnames(ge.exp2) <- samples[samples_ge, ]$sample_id
      colnames(cn2) <- samples[samples_cn, ]$sample_id
      colnames(me.exp2) <- samples[samples_me, ]$sample_id
      
      ## removing genes with zero and near zero variance in gene expression dataset and creating final dataset
      # ge <-  Gene expression dataframe
      # cnv <- copy number variation dataframe
      # me <- methylation intensity dataframe
      exp_var <- rowVars(as.matrix(log2(ge.exp2 + 1)))
      data_merged.ge <- ge.exp2[exp_var > 0.25, ]
      data_merged.ge <- data.frame(data_merged.ge)
      
      data_merged.cnv <- cn2[rownames(data_merged.ge), ]
      data_merged.me <- me.exp2[rownames(data_merged.ge), ]
      data_merged.samples <- samples
      
      message("Saving merged datasets to:", file.path(data_folder, paste0(data_name, "_RawDataset.Rdata")))
      save(file = file.path(data_folder, paste0(data_name, "_RawDataset.Rdata")), data_merged.ge, data_merged.cnv, data_merged.me, data_merged.samples)
}


```
## Dataset generation

Datasets were downloaded on 25/06/2021 by Ariane Mora.

Note you need to unzip the phenotype files.

### KIRC

0) Phenotype: https://gdc-hub.s3.us-east-1.amazonaws.com/download/TCGA-KIRC.GDC_phenotype.tsv.gz
1) CNV: TCGA-KIRP.gistic.tsv
    1.a) copy number (gene-level) - GISTIC - focal score by gene  
    1.b) https://gdc-hub.s3.us-east-1.amazonaws.com/download/TCGA-KIRC.gistic.tsv.gz
2) DNA Methylation: TCGA-KIRP.methylation450.tsv: 
    2.a) DNA methylation - Illumina Human Methylation 450          
    2.b) https://gdc-hub.s3.us-east-1.amazonaws.com/download/TCGA-KIRC.methylation450.tsv.gz
3) RNAseq counts: TCGA-SKCM.htseq_counts.tsv
    3.a) gene expression RNAseq - HTSeq - Counts unit: log2(count+1)
    3.b) https://gdc-hub.s3.us-east-1.amazonaws.com/download/TCGA-KIRC.htseq_counts.tsv.gz


                       
                       
### KIRP
https://xenabrowser.net/datapages/?cohort=GDC%20TCGA%20Kidney%20Papillary%20Cell%20Carcinoma%20(KIRP)&removeHub=https%3A%2F%2Fxena.treehouse.gi.ucsc.edu%3A443
0) Phenotype: https://gdc-hub.s3.us-east-1.amazonaws.com/download/TCGA-KIRP.GDC_phenotype.tsv.gz
1) CNV: TCGA-KIRP.gistic.tsv
    1.a) copy number (gene-level) - GISTIC - focal score by gene  
    1.b) https://gdc-hub.s3.us-east-1.amazonaws.com/download/TCGA-KIRP.cnv.tsv.gz
2) DNA Methylation: TCGA-KIRP.methylation450.tsv: 
    2.a) DNA methylation - Illumina Human Methylation 450          
    2.b) https://gdc-hub.s3.us-east-1.amazonaws.com/download/TCGA-KIRP.methylation450.tsv.gz
3) RNAseq counts: TCGA-SKCM.htseq_counts.tsv
    3.a) gene expression RNAseq - HTSeq - Counts unit: log2(count+1)
    3.b) https://gdc-hub.s3.us-east-1.amazonaws.com/download/TCGA-KIRP.htseq_counts.tsv.gz
    
    Additional - New Primary            Primary Tumor      Solid Tissue Normal 
                           1                      292                       88
                           
### Dataset download for Melanoma 
     
#### Main site:  https://xenabrowser.net/datapages/?cohort=GDC%20TCGA%20Melanoma%20(SKCM)&removeHub=https%3A%2F%2Fxena.treehouse.gi.ucsc.edu%3A443  
0) Phenotype: https://gdc-hub.s3.us-east-1.amazonaws.com/download/TCGA-SKCM.GDC_phenotype.tsv.gz
1) CNV: TCGA-SKCM.gistic.tsv
    1.a) copy number (gene-level) - GISTIC - focal score by gene  
    1.b) https://gdc-hub.s3.us-east-1.amazonaws.com/download/TCGA-SKCM.gistic.tsv.gz
2) DNA Methylation: TCGA-SKCM.methylation450.tsv: 
    2.a) DNA methylation - Illumina Human Methylation 450          
    2.b) https://gdc-hub.s3.us-east-1.amazonaws.com/download/TCGA-SKCM.methylation450.tsv.gz
3) RNAseq counts: TCGA-SKCM.htseq_counts.tsv
    3.a) gene expression RNAseq - HTSeq - Counts unit: log2(count+1)
    3.b) https://gdc-hub.s3.us-east-1.amazonaws.com/download/TCGA-SKCM.htseq_counts.tsv.gz

    Additional Metastatic            Metastatic         Primary Tumor   Solid Tissue Normal 
                        2                   368                   104                     3 

```{r}
make_datasets("PAAD", ".")
make_datasets("COAD", ".")
make_datasets("KIRP", ".")
make_datasets("KIRC", ".")
make_datasets("SKCM", ".")
make_datasets("MESO", ".")

load(file.path('.', paste0("PAAD", "_RawDataset.Rdata")))

```

### Dataset download for Mesothelioma
     
#### Main site: https://xenabrowser.net/datapages/?cohort=GDC%20TCGA%20Mesothelioma%20(MESO)&removeHub=https%3A%2F%2Fxena.treehouse.gi.ucsc.edu%3A443
0) Phenotype: https://gdc-hub.s3.us-east-1.amazonaws.com/download/TCGA-MESO.GDC_phenotype.tsv.gz
1) CNV: TCGA-MESO.gistic.tsv
    1.a) copy number (gene-level) - GISTIC - focal score by gene  
    1.b) https://gdc-hub.s3.us-east-1.amazonaws.com/download/TCGA-MESO.gistic.tsv.gz
2) DNA Metylation: TCGA-MESO.methylation450.tsv  
    2.a) DNA methylation - Illumina Human Methylation 450  
    2.b) https://gdc-hub.s3.us-east-1.amazonaws.com/download/TCGA-MESO.methylation450.tsv.gz
3) RNAseq counts: TCGA-MESO.htseq_counts.tsv
    3.a) gene expression RNAseq - HTSeq - Counts log2(count+1)
    3.b) https://gdc-hub.s3.us-east-1.amazonaws.com/download/TCGA-MESO.htseq_counts.tsv.gz

      Primary Tumor Solid Tissue Normal 
                 87                   1 

OMITTED since there were insufficient normal tissues
```{r}
make_datasets("MESO", "MESO")
```

### Dataset download for Colon
     
#### Main site: https://xenabrowser.net/datapages/?cohort=GDC%20TCGA%20Colon%20Cancer%20(COAD)
0) Phenotype: https://gdc-hub.s3.us-east-1.amazonaws.com/download/TCGA-COAD.GDC_phenotype.tsv.gz
1) CNV: TCGA-COAD.gistic.tsv
    1.a) copy number (gene-level) - GISTIC - focal score by gene unit Gistic2 copy number  
    1.b) https://gdc-hub.s3.us-east-1.amazonaws.com/download/TCGA-COAD.gistic.tsv.gz
2) DNA Metylation: TCGA-MESO.methylation450.tsv  
    2.a) DNA methylation - Illumina Human Methylation 450  unit: beta value
    2.b) https://gdc-hub.s3.us-east-1.amazonaws.com/download/TCGA-COAD.methylation450.tsv.gz
3) RNAseq counts: TCGA-COAD.htseq_counts.tsv
    3.a) gene expression RNAseq - HTSeq - Counts unit: log2(count+1)
    3.b) https://gdc-hub.s3.us-east-1.amazonaws.com/download/TCGA-COAD.methylation450.tsv.gz
    
         Metastatic       Primary Tumor     Recurrent Tumor Solid Tissue Normal 
                  2                 475                   1                  93 
```{r}
make_datasets("COAD", "COAD")
```

#### Main site: https://xenabrowser.net/datapages/?cohort=GDC%20TCGA%20Pancreatic%20Cancer%20(PAAD)&removeHub=https%3A%2F%2Fxena.treehouse.gi.ucsc.edu%3A443
0) Phenotype: https://gdc-hub.s3.us-east-1.amazonaws.com/download/TCGA-PAAD.GDC_phenotype.tsv.gz
1) CNV: TCGA-PAAD.gistic.tsv
    1.a) copy number (gene-level) - GISTIC - focal score by gene unit Gistic2 copy number  
    1.b) https://gdc-hub.s3.us-east-1.amazonaws.com/download/TCGA-PAAD.gistic.tsv.gz
2) DNA Metylation: TCGA-PAAD.methylation450.tsv
    2.a) DNA methylation - Illumina Human Methylation 450  unit: beta value
    2.b) https://gdc-hub.s3.us-east-1.amazonaws.com/download/TCGA-PAAD.methylation450.tsv.gz
3) RNAseq counts: TCGA-PAAD.htseq_counts.tsv
    3.a) gene expression RNAseq - HTSeq - Counts unit: log2(count+1)
    3.b) https://gdc-hub.s3.us-east-1.amazonaws.com/download/TCGA-PAAD.htseq_counts.tsv.gz
    

         Metastatic       Primary Tumor Solid Tissue Normal 
                  1                 185                  37 
```{r}
make_datasets("PAAD", "PAAD")
```