---
title: "Multi-stage tool analysis"
output: html_notebook
---

# Install and setup

This is an [R Markdown](http://rmarkdown.rstudio.com) Notebook. When you execute code within the notebook, the results appear beneath the code. 

Try executing this chunk by clicking the *Run* button within the chunk or by placing your cursor inside it and pressing *Cmd+Shift+Enter*. 

```{r}
MethylMix_ModelGeneExpression_FDR <- function (METcancer, GEcancer, CovariateData = NULL) 
{
  OverlapSamples = intersect(colnames(METcancer), colnames(GEcancer))
  cat("Found", length(OverlapSamples), "samples with both methylation and expression data.\n")
  GEcancer = GEcancer[, OverlapSamples, drop = FALSE]
  METcancer = METcancer[, OverlapSamples, drop = FALSE]
  if (!is.null(CovariateData)) 
    CovariateData = CovariateData[OverlapSamples, , drop = FALSE]
  Rsquares = matrix(0, nrow = length(rownames(METcancer)), 
                    ncol = 1)
  Genes = rownames(METcancer)
  PvalueThreshold = 0.001
  RsquareThreshold = 0.1
  cat("Correlating methylation data with gene expression...\n")
  i <- NULL
  Rsquares = foreach::foreach(i = 1:length(rownames(METcancer)), 
                              .combine = "c") %dopar% {
                                Rsq = c(0,0)
                                tmpGene = unlist(strsplit(Genes[i], "---"))[1]
                                pos = which(rownames(GEcancer) == tmpGene)
                                if (length(pos) > 0) {
                                  if (!is.null(CovariateData)) {
                                    res = lm(GEcancer[pos, ] ~ METcancer[Genes[i], ] + factor(CovariateData))
                                    res.summary = summary(res)
                                    an = anova(res)
                                    if (res$coefficients[2] < 0 & res.summary$coefficients[2,4] < PvalueThreshold) {
                                      Rsq = res.summary$r.squared
                                    }
                                  }
                                  else {
                                    res = lm(GEcancer[pos, ] ~ METcancer[Genes[i], ])
                                    res.summary = summary(res)
                                    if (res$coefficients[2] < 0 & res.summary$coefficients[2,4] < PvalueThreshold) {
                                      Rsq = c(res.summary$r.squared, res.summary$coefficients[2,4])
                                    }
                                  }
                                }
                                Rsq
                              }
  Rsquares = data.frame(matrix(Rsquares, ncol = 2, byrow = T))
  Rsquares[,3] = p.adjust(Rsquares[,2], method = "fdr", n = length(Rsquares[,2])) # estimating fdr
  FunctionalGenes = Genes[Rsquares[,1] > RsquareThreshold]
  cat("\nFound", length(FunctionalGenes), "transcriptionally predictive genes.\n")
  
  Rsquares = Rsquares[Rsquares[,1] > RsquareThreshold,]
  FunctionalGenes_withfdr = cbind(FunctionalGenes, Rsquares[,3])
  return(FunctionalGenes_withfdr)
}


```

## CNAmet

Install CNAmet from: https://csbi.ltdk.helsinki.fi/CNAmet/

```{r}
BiocManager::install("CNAmet")
library(CNAmet)
library(tictoc)

data_name <- 'PAAD'
data_folder <- '.'
# load data
load(file.path(data_folder, paste0(data_name, "_RawDataset.Rdata")))

### reading copy number values and creating two binary matrices for amplification and deletion calls
cn.amp <- data_merged.cnv
cn.amp[cn.amp > 1] <- 1
cn.amp[cn.amp < 0] <- 0

cn.del <- data_merged.cnv
cn.del[cn.del > 0] <- 0
cn.del[cn.del < 0] <- 1
cn.del[1:5,1:5]

cn.amp <- as.matrix(cn.amp, rownames.force = T)
cn.del <- as.matrix(cn.del, rownames.force = T)
ge.exp <- as.matrix(data_merged.ge, rownames.force = T)
me.m <- as.matrix(data_merged.me, rownames.force = T) # Convert to M values as recomended by their how to: https://csbi.ltdk.helsinki.fi/CNAmet/howto.html
me.m <- log2(me.m/(1 - me.m)) # Formula for converting Beta values to M values.

# running CNAmet
tic("Total")

tic("Amplification")
amp_results <- CNAmet(exprMatrix = ge.exp, cghMatrix = cn.amp, methylMatrix = me.m, perms = 1000, na.limit = 0.1, gainData = TRUE,
                     favorSynergetic = TRUE, strictChecks = FALSE, strictLim = 0.05)
toc()

tic("Deletion")
del_results <- CNAmet(exprMatrix = ge.exp, cghMatrix = cn.del, methylMatrix = me.m, perms = 1000, na.limit = 0.1, gainData = FALSE,
                     favorSynergetic = TRUE, strictChecks = FALSE, strictLim = 0.05)
toc()
toc()
# Amplification: 2454.671 sec elapsed
# Deletion: 2195.002 sec elapsed
# Total: 4649.674 sec elapsed

amp_results <- data.frame(amp_results)
del_results <- data.frame(del_results)
amp_results$Gene <- rownames(amp_results)
del_results$Gene <- rownames(del_results)

# saving results
write.table(amp_results, file = paste0("CNAmet_", data_name, "_GainDriven.tsv"), row.names = F, sep = "\t", quote = F)
write.table(del_results, file = paste0("CNAmet_", data_name, "_LossDriven_Genes.tsv"), row.names = F, sep = "\t", quote = F)

```


## MethylMix

```{r}
library(MethylMix)
library(tictoc)

data_name <- 'PAAD'
data_folder <- '.'
# load data
load(file.path(data_folder, paste0(data_name, "_RawDataset.Rdata")))


# load data
gene_exp <- as.matrix(data_merged.ge, rownames.force = T)
meth_beta <- as.matrix(data_merged.me, rownames.force = T)
samples <- data_merged.samples

# Need to get just the cancer samples for the gene expression 
gene_exp_cancer <- gene_exp[, grepl("Tumor", colnames(gene_exp))]
meth_beta_cancer <- meth_beta[, grepl("Tumor", colnames(meth_beta))]
meth_beta_normal <- meth_beta[, grepl("Normal", colnames(meth_beta))]

# running methylmix in parallel
tic("MethylMix")
methyl_mix_genes <- MethylMix(meth_beta_cancer, gene_exp_cancer, meth_beta_normal)
toc()
# Non paralell run: 29/06/2021
# MethylMix: 975.421 sec elapsed

genes <- as.character(methyl_mix_genes)

# saving results
write.table(genes, paste0("MethylMix_", data_name, "_3000Genes.txt"), quote = F, row.names = F)

```

## EpiMethEx
https://bmcbioinformatics.biomedcentral.com/articles/10.1186/s12859-018-2397-6
https://github.com/giupardeb/EpiMethEx

* checking for file ‘EpiMethEx/DESCRIPTION’ ... OK
* preparing ‘EpiMethEx’:
* checking DESCRIPTION meta-information ... OK
* installing the package to build vignettes
      -----------------------------------
ERROR: dependencies ‘miscTools’, ‘MultiAssayExperiment’ are not available for package ‘EpiMethEx’
* removing ‘/private/var/folders/gq/6ljhmvm1713fykdjqbl188pm0000gn/T/Rtmpv4Ei4J/Rinstfcf6ed3ee51/EpiMethEx’
      -----------------------------------
ERROR: package installation failed

Had to install the required packages:
BiocManager::install("MultiAssayExperiment")
install.packages("miscTools")


* checking for file ‘EpiMethEx/DESCRIPTION’ ... OK
* preparing ‘EpiMethEx’:
* checking DESCRIPTION meta-information ... OK
* installing the package to build vignettes
* creating vignettes ... ERROR
--- re-building ‘epimethex_analysis.Rmd’ using rmarkdown
Warning in xtfrm.data.frame(x) : cannot xtfrm data frames
Warning in xtfrm.data.frame(x) : cannot xtfrm data frames
Warning in xtfrm.data.frame(x) : cannot xtfrm data frames
Quitting from lines 44-80 (epimethex_analysis.Rmd)
Error: processing vignette 'epimethex_analysis.Rmd' failed with diagnostics:
there is no package called 'RcmdrMisc'
--- failed re-building ‘epimethex_analysis.Rmd’

SUMMARY: processing the following file failed:
  ‘epimethex_analysis.Rmd’

Error: Vignette re-building failed.
Execution halted

Had to then do:
BiocManager::install("MultiAssayExperiment")
install.packages("miscTools")
install.packages("RcmdrMisc")
install.packages("psych")

Still unable to work...
* checking for file ‘EpiMethEx/DESCRIPTION’ ... OK
* preparing ‘EpiMethEx’:
* checking DESCRIPTION meta-information ... OK
* installing the package to build vignettes
* creating vignettes ... ERROR
--- re-building ‘epimethex_analysis.Rmd’ using rmarkdown
Warning in xtfrm.data.frame(x) : cannot xtfrm data frames
Warning in xtfrm.data.frame(x) : cannot xtfrm data frames
Warning in xtfrm.data.frame(x) : cannot xtfrm data frames
Warning in `[<-.factor`(`*tmp*`, ri, value = -1.859692469761) :
  invalid factor level, NA generated
Warning in `[<-.factor`(`*tmp*`, ri, value = -0.396742484881) :
  invalid factor level, NA generated
Warning in `[<-.factor`(`*tmp*`, ri, value = 0.225157530238) :
  invalid factor level, NA generated
Quitting from lines 44-80 (epimethex_analysis.Rmd)
Error: processing vignette 'epimethex_analysis.Rmd' failed with diagnostics:
task 1 failed - "undefined columns selected"
--- failed re-building ‘epimethex_analysis.Rmd’

SUMMARY: processing the following file failed:
  ‘epimethex_analysis.Rmd’

Error: Vignette re-building failed.
Execution halted

library(EpiMethEx)

gene_exp <- as.matrix(data_merged.ge, rownames.force = T)
meth_beta <- as.matrix(data_merged.me, rownames.force = T)
meth_annotations <- read.csv('illuminaMethyl450_hg38_GDC_2500_promoter_mapping.csv')
epimethex.analysis(Expressions, Annotations, Methylation, 1, 3, 2,TRUE, TRUE, FALSE)


## MEAL
https://bioconductor.org/packages/3.13/bioc/html/MEAL.html
https://www.bioconductor.org/packages/release/bioc/vignettes/MEAL/inst/doc/caseExample.html#4_Methylation_and_gene_expression_integration

# Had to install 
```{r}
library("MEAL")
library("minfi")

multi <- createMultiDataSet()
annotation <- read.csv('tcga_hsapiens_gene_ensembl-GRCh38.p13.csv')
annotation <- annotation[!is.na(annotation$external_gene_name), ]
annotation <- annotation[!is.na(annotation$start_position), ]
annotation <- annotation[order(annotation$ensembl_gene_id, annotation$entrezgene_id), ]
annotation <- annotation[!duplicated(annotation$ensembl_gene_id),]
rownames(annotation) <- annotation$ensembl_gene_id
gene_exp <- as.matrix(data_merged.ge, rownames.force = T)
meth_beta <- as.matrix(data_merged.me, rownames.force = T)
meth_annot <- AnnotatedDataFrame(me.annot)
rownames(meth_beta) <- meth_annot$id # Change to be CpG IDs for this one.

# have to add "chromosome" and "start", "end" for the eset
samples <- data_merged.samples
rownames(samples) <- samples$sample_id

# Create the phenotype data frame
gene_exp_pdata <- samples[colnames(gene_exp), ]
# create the feature data by intersecting with the annotation file
annot_genes <- intersect(rownames(annotation), rownames(gene_exp))
gene_exp <- gene_exp[annot_genes, ]
gene_annot <- annotation[annot_genes, ]
gene_exp_f_data <- data.frame(chromosome = as.character(gene_annot$chromosome_name), 
                              start = as.numeric(gene_annot$start_position),
                              end = as.numeric(gene_annot$end_position), 
                              stringsAsFactors = FALSE)
rownames(gene_exp_f_data) <- annot_genes

# Need to get just the cancer samples for the gene expression 
gene_exp_eset <- ExpressionSet(assayData = gene_exp,
                               phenoData = AnnotatedDataFrame(gene_exp_pdata), 
                               featureData = AnnotatedDataFrame(gene_exp_f_data))

multi <- add_genexp(multi, gene_exp_eset)
# Do the same for the beta data
gene_exp_pdata <- AnnotatedDataFrame[colnames(gene_exp), ]
meth_grset <- makeGenomicRatioSetFromMatrix(meth_beta, rownames = rownames(meth_beta), pData = meth_pdata,
                              array = "IlluminaHumanMethylation450k",
                              mergeManifest = FALSE, what = c("Beta"))

multi <- add_methy(multi, meth_beta)
methExprs <- correlationMethExprs(multi)
head(methExprs)

```

## BioMethyl

Notes:
unable to install from bioconductor so had to install from source
then didn't run because it was trying to install packages using bioclite
had to install ENmix using bioconductor, then finall got error: Error: 'rm.outlier' is not an exported object from 'namespace:ENmix'
Thus abandoned.
https://academic.oup.com/bioinformatics/article/35/19/3635/5364266
https://github.com/yuewangpanda/BioMethyl

Error: 'rm.outlier' is not an exported object from 'namespace:ENmix'

Downloaded package from: https://github.com/yuewangpanda/BioMethyl/blob/master/BioMethyl_1.1.tar.gz
```{r}
# install.packages("BioMethyl_1.1.tar.gz", repos = NULL)
library("BioMethyl")
#BiocManager::install("ENmix")
gene_exp <- as.matrix(data_merged.ge, rownames.force = T)
meth_beta <- as.matrix(data_merged.me, rownames.force = T)

dat <- filterMethyData(meth_beta)
myExpr <- calExpr(dat, "BRCA", Example = T, SaveOut = F, "")
samp_1 <- colnames(myExpr)[1:10]
samp_2 <- colnames(myExpr)[11:20]
mydf <- calDEG(myExpr, Sample_1 = samp_1, Sample_2 = samp_2, SaveOut = F, "")

```
## SIRCLE

```{r}

```